exercise/cache/points.py

```diff
@@ -73,6 +76,68 @@ def _generate_data(
             user: User,
             data: Optional[Dict[str, Any]] = None,
             ) -> Dict[str, Any]:
+        # Perform all database queries before generating the cache.
+        exercises = list(
```

Comment:
The memory consumption is a good point. I didn't consider that, and instead only optimized for minimal execution time and minimal number of database queries. The generation code seemed to work ever so slightly faster when using `list()`, so I opted to use it. In addition, I logged all points when the generation code executes a query, and it seemed that when not using `list()`, some queries were executed during the second pass (when `is_staff=False`) because of reveal rules. I thought it might be more efficient to do all database queries upfront.

>I was under the impression that QuerySets cache the results inside the QuerySet object and you could iterate over the same QuerySet twice while loading the data only once from the database. I am not sure, though.

>Then again, is the QuerySet's cache actually in memory, thus either the memory is consumed by this list variable or a QuerySet variable?

Yes, the results are cached in the `QuerySet` object, so I would imagine either the list or the queryset would consume the memory. I'm not sure if it's the same amount of memory, though, since I didn't measure. It might be that the queryset's cache is somehow more optimized?

Reply:
According to your test, the QuerySet object consumes even a little more memory than the simple list. Let's keep the list as is.